readme.md
[[!meta title="dropbox"]]

* Authors: Patrick ZAJDA, Filaos and other contributors
* NVDA compatibility: 2019.1 or later

This plugin add a shortcut to announce Dropbox status or open the Dropbox systray menu when pressed once or twice respectively.

* Shortcut: NVDA+Alt+D


## Changes for 4.7 ##

* Compatibility with NVDA 2022.1
* Remove Dropbox app module as it causes issues with last Dropbox interface and Windows 8 is not supported by Microsoft

## Changes for 4.6 ##

* Specify NVDA 2021.1 compatibility

## Changes for 4.4 ##

* Python 3 compatibility
* Use the last addon template
* Repository change to be built with Appveyor
* Fixed wrong and removed unused shortcuts in the documentation
* Update the description in the documentation which still referenced the announcement of the version

## Changes for 4.0 ##

* Change Shift+NVDA+D behavior: if pressed once announce Dropbox status, twice open the context menu.
* Translations update.
* Fixed issues with windows 8 metro app.
* Addon help is no longer available in the help menu, it's found in the addons manager instead.
* As Dropbox changes his gui from version to version, all features for settings dialog were removed.
* Modified the shortcut to get dropbox status to Alt+NVDA+D to avoid conflict with audio ducking support.

## Changes for 3.1 ##

* Use another way to get cancel button and page tab. Now we don't have to focus them before using shortcuts.
* When changing the active tab, the focus move to the tab page so when pressing tab, the first item of the tab is activated instead of staying to the previous used tab even if it is not activated anymore.
* In the preferences dialog, it is now possible to press control+page up/down to switch between tabs. Control+tab and control+shift+tab still work.
* All localized manifest files should now be OK.
* Minor corrections.

## Changes for 3.0 ##

* Minor correction in the main manifest file (authors are correctly displayed).
* Improved context menu detection when pressing Shift+NVDA+D three times.
* The escape button now works (only when using Dropbox in the same language NVDA uses).
* A lot of corrections in the code.
* Added/updated documentations of all scripts.
* New languages: Arabic, Brazilian Portuguese, Czech, Dutch, Finnish, Galician, German, Hungarian, Japanese, Nepali, Polish, Russian, Spanish, Slovak, Tamil, Turkish.

## Changes for 2.0 ##

* New languages: Italian
* Pressing the shortcut three times or more when already being in the context menu doesn't cause problem anymore.

## Changes for 1.0 ##

* Initial Release



changelog.md
## Changes for 2023.3 ##

* Compatibility with NVDA 2023.1
* New yyyy.m version style
* Translations update



addon/appModules/dropbox.py
# -*- coding: utf-8 -*-
# Dropbox appModule for NVDA
# Improves accessibility of the Dropbox Metro app for Windows 8
# Copyright (C) 2015 Filaos, Patrick ZAJDA <patrick@zajda.fr> and other contributors
# Windows 8 compatibility contributed by David Parduhn <xkill85@gmx.net>
# This file is covered by the GNU General Public License.
# You can read the licence by clicking Help->Licence in the NVDA menu
# or by visiting http://www.gnu.org/licenses/old-licenses/gpl-2.0.html

import addonHandler
import appModuleHandler
from NVDAObjects import UIA
import controlTypes

# We keep it in the case it would finally be necessary to change add-on summary by language
# _addonDir = os.path.join(os.path.dirname(__file__), "..").decode("mbcs")
# _curAddon = addonHandler.Addon(_addonDir)
# _addonSummary = _curAddon.manifest['summary']

# We initialize translations
addonHandler.initTranslation()
# We define roles to keep compatibility with nvda<2022.1
if hasattr(controlTypes, 'ROLE_STATICTEXT'):
	STATIC_TEXT = controlTypes.ROLE_STATICTEXT
else:
	STATIC_TEXT = controlTypes.Role.STATICTEXT
if hasattr(controlTypes, 'ROLE_LISTITEM'):
	LIST_ITEM = controlTypes.ROLE_LISTITEM
else:
	LIST_ITEM = controlTypes.Role.LISTITEM


# Class for Dropbox items in the metro app
class dropboxitem(UIA. ListItem):
	def _get_name(self):
		dbxList = list()
		obj = self.firstChild
		if obj.name != u'':
			dbxList.append(obj.name)
		while (obj != self.lastChild):
			if obj.role == STATIC_TEXT:
				if obj.name != u'':
					dbxList.append(obj.name)
			obj = obj.next
		return '; '.join(dbxList)


class AppModule(appModuleHandler.AppModule):
	# We set the scripts category shown on input gestures dialog
	# scriptCategory = unicode(_addonSummary)
	scriptCategory = u"Dropbox"

	def chooseNVDAObjectOverlayClasses(self, obj, CLSList):
		if obj.role == LIST_ITEM:
			CLSList.insert(0, dropboxitem)


addon/globalPlugins/SayStatedropbox.py
# -*- coding: utf-8 -*-
# Dropbox announcement Global Plugin for NVDA
# Copyright (C) 2014 Filaos, Patrick ZAJDA <patrick@zajda.fr>, Rui Fontes and other contributors
# This file is covered by the GNU General Public License.
# You can read the licence by clicking Help->Licence in the NVDA menu
# or by visiting http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
# Shortcut: NVDA+Alt+D
# Some portion have been directly copied from the systraylist addon
# copyright (C) Rui Fontes and other contributors

import ctypes
import globalPluginHandler
import addonHandler
import scriptHandler
import ui
import NVDAObjects
import api
import windowUtils
import winUser
import controlTypes
from typing import Callable


# We initialize translation support
addonHandler.initTranslation()
_: Callable[[str], str]
# Constents for roles to keep compatibility
if hasattr(controlTypes, 'ROLE_POPUPMENU'):
	POPUP_MENU = controlTypes.ROLE_POPUPMENU
else:
	POPUP_MENU = controlTypes.Role.POPUPMENU
if hasattr(controlTypes, 'ROLE_MENUITEM'):
	MENU_ITEM = controlTypes.ROLE_MENUITEM
else:
	POPUP_MENU = controlTypes.Role.MENUITEM


class GlobalPlugin(globalPluginHandler.GlobalPlugin):
	# We initialize the scripts category shown on input gestures dialog
	scriptCategory = u"Dropbox"

	def _findAccessibleLeafsFromWindowClassPath(self, windowClassPath):
		# Create a list of systray icons
		# Copied from systray list add-on from Rui Fontes
		h, FindWindowEx = 0, winUser.user32.FindWindowExW
		for element in windowClassPath:
			h = FindWindowEx(h, 0, element, 0)
		trayObjects = []
		trayObject = NVDAObjects.IAccessible.getNVDAObjectFromEvent(h, -4, 1)
		# When trayObject.next is None it means that there is no more objects on the systray.
		while trayObject is not None:
			if trayObject.name:
				trayObjects.append(trayObject)
			trayObject = trayObject.next
		return trayObjects

	def _findAccessibleLeafsFromWindowClassPath11(self, windowClassPath):
		# Create a list of systray icons
		# Copied from Systray list NVDA add-on
		h = 0
		for className in windowClassPath:
			h = ctypes.windll.user32.FindWindowExA(h, 0, className, 0)
			# if not h:
			# 	break
		obj = NVDAObjects.IAccessible.getNVDAObjectFromEvent(h, -4, 0).firstChild.children
		trayObjects = []
		for trayObject in obj:
			trayObjects.append(trayObject)
		return trayObjects

	def _findAccessibleLeafsFromWindowClassPath11_22h2(self):
		"""
		Create a list of systray icons
		Copied from Systray list add-on
		"""
		# Starting to find the handle of the Shell_TrayWnd window
		h = winUser.FindWindow("Shell_TrayWnd", None)
		# Now, lets get the handle of Windows.UI.Input.InputSite.WindowClass window, where the icons reside...
		hwnd = windowUtils.findDescendantWindow(
			h, visible=None, controlID=None, className="Windows.UI.Input.InputSite.WindowClass"
		)
		# Now, lets get all objects in this window and its location
		obj = NVDAObjects.IAccessible.getNVDAObjectFromEvent(hwnd, -4, 0).children
		trayObjects = []
		# We start in the second object because the first object do not interesse us...
		o = 1
		while o in range(len(obj)):
			trayObjects.append(obj[o])
			o = o + 1
		return trayObjects

	def _findDropbox(self):
		path = ("shell_TrayWnd", "TrayNotifyWnd", "SysPager", "ToolbarWindow32")
		path11 = ("shell_TrayWnd", "TrayNotifyWnd", "Windows.UI.Composition.DesktopWindowContentBridge")
		try:
			from winVersion import getWinVer, WinVersion
			global win11_22h2
			win11_22h2 = getWinVer() >= WinVersion(major=10, minor=0, build=22621)
			win11 = getWinVer() >= WinVersion(major=10, minor=0, build=22000)
		except ImportError:
			win11 = False
		if win11_22h2:
			objects = self._findAccessibleLeafsFromWindowClassPath11_22h2()
		elif win11:
			objects = self._findAccessibleLeafsFromWindowClassPath11(path11)
		else:
			objects = self._findAccessibleLeafsFromWindowClassPath(path)
		for trayIcon in objects:
			iconName = trayIcon.name
			if iconName is not None and "dropbox" in iconName.lower():
				# dropbox object found
				return trayIcon
		return None

	def script_announceDropbox(self, gesture):
		trayIcon = self._findDropbox()
		if not trayIcon:
			# Translators: the message presented when Dropbox tray icon was not found
			ui.message(_("drop box not found"))
			return

		name = trayIcon.name.split()
		repeatCount = scriptHandler.getLastScriptRepeatCount()
		if repeatCount == 0:
			# announce dropbox state
			if win11_22h2:
				del(name[0:3])
			del(name[1])
			name = " ".join(name)
			ui.message(name)

		else:
			# activate dropbox icon
			# If we are already inside of the context menu, stop the script
			objFocused = api.getFocusObject()
			currentProcess = objFocused.appModule.appName.lower()
			if (
				(currentProcess.lower() == u'dropbox' and objFocused.windowClassName.lower() == u'#32768')
				and objFocused.role in ({POPUP_MENU, MENU_ITEM})
			):
				return
			else:
				try:
					trayIcon.doAction()
				except NotImplementedError:
					# Translators: the message reported when it is not possible to activate Dropbox context menu.
					ui.message(_("Unable to activate Dropbox context menu"))
	# Translators: message presented when user performes input help for the Shift+NVDA+D script
	script_announceDropbox.__doc__ = _(
		"If pressed once, announces Dropbox status. "
		"If pressed twice, open the Dropbox context menu by clicking on its systray icon"
	)

	__gestures = {
		"kb:NVDA+alt+d": "announceDropbox",
	}


buildVars.py
# -*- coding: UTF-8 -*-

# Build customizations
# Change this file instead of sconstruct, whenever possible.

# Since some strings in `addon_info` are translatable,
# we need to include them in the .po files.
# Gettext recognizes only strings given as parameters to the `_` function.
# To avoid initializing translations in this module we simply roll our own "fake" `_` function
# which returns whatever is given to it as an argument.
def _(arg):
	return arg


# Add-on information variables
addon_info = {
	# add-on Name
	"addon_name" : "dropbox",
	# Add-on summary, usually the user visible name of the addon.
	# Translators: Summary for this add-on to be shown on installation and add-on information.
	"addon_summary" : "Dropbox",
	# Add-on description
	# Translators: Long description to be shown for this add-on on installation and add-on information
	"addon_description" : _("""Announces Dropbox status or open the Dropbox systray menu.
Shortcut: NVDA+Alt+D"""),
	# version
	"addon_version" : "2023.3",
	# Author(s)
	"addon_author" : "Patrick ZAJDA <patrick@zajda.fr>, Filaos and other contributors",
	# URL for the add-on documentation support
	"addon_url" : "https://addons.nvda-project.org/",
	# URL for the add-on repository where the source code can be found
	"addon_sourceURL": "https://github.com/Nardol/dropbox",
	# Documentation file name
	"addon_docFileName" : "readme.html",
	# Minimum NVDA version supported (e.g. "2018.3.0", minor version is optional)
	"addon_minimumNVDAVersion" : "2019.1",
	# Last NVDA version supported/tested (e.g. "2018.4.0", ideally more recent than minimum version)
	"addon_lastTestedNVDAVersion" : "2023.1",
	# Add-on update channel (default is stable or None)
	"addon_updateChannel" : None,
	# Add-on license such as GPL 2
	"addon_license": "GPL V2",
	# URL for the license document the ad-on is licensed under
	"addon_licenseURL": "https://www.gnu.org/licenses/old-licenses/gpl-2.0.html",
}


import os.path

# Define the python files that are the sources of your add-on.
# You can use glob expressions here, they will be expanded.
pythonSources = [os.path.join("addon", "globalPlugins", "*.py")]

# Files that contain strings for translation. Usually your python sources
i18nSources = pythonSources + ["buildVars.py"]

# Files that will be ignored when building the nvda-addon file
# Paths are relative to the addon directory, not to the root directory of your addon sources.
excludedFiles = []

# Base language for the NVDA add-on
# If your add-on is written in a language other than english, modify this variable.
# For example, set baseLanguage to "es" if your add-on is primarily written in spanish.
baseLanguage = "en"

# Markdown extensions for add-on documentation
# Most add-ons do not require additional Markdown extensions.
# If you need to add support for markup such as tables, fill out the below list.
# Extensions string must be of the form "markdown.extensions.extensionName"
# e.g. "markdown.extensions.tables" to add tables.
markdownExtensions = []
